File: benchmarks3d/joint_revolute3.rs
use rapier_testbed3d::Testbed;
use rapier3d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut bodies = RigidBodySet::new();
    let mut colliders = ColliderSet::new();
    let mut impulse_joints = ImpulseJointSet::new();
    let multibody_joints = MultibodyJointSet::new();

    let rad = 0.4;
    let num = 10;
    let shift = 2.0;

    for l in 0..4 {
        let y = l as f32 * shift * (num as f32) * 3.0;

        for j in 0..50 {
            let x = j as f32 * shift * 4.0;

            let ground = RigidBodyBuilder::fixed().translation(vector![x, y, 0.0]);
            let mut curr_parent = bodies.insert(ground);
            let collider = ColliderBuilder::cuboid(rad, rad, rad);
            colliders.insert_with_parent(collider, curr_parent, &mut bodies);

            for i in 0..num {
                // Create four bodies.
                let z = i as f32 * shift * 2.0 + shift;
                let positions = [
                    Isometry::translation(x, y, z),
                    Isometry::translation(x + shift, y, z),
                    Isometry::translation(x + shift, y, z + shift),
                    Isometry::translation(x, y, z + shift),
                ];

                let mut handles = [curr_parent; 4];
                for k in 0..4 {
                    let density = 1.0;
                    let rigid_body = RigidBodyBuilder::dynamic().position(positions[k]);
                    handles[k] = bodies.insert(rigid_body);
                    let collider = ColliderBuilder::cuboid(rad, rad, rad).density(density);
                    colliders.insert_with_parent(collider, handles[k], &mut bodies);
                }

                // Setup four impulse_joints.
                let x = Vector::x_axis();
                let z = Vector::z_axis();

                let revs = [
                    RevoluteJointBuilder::new(z).local_anchor2(point![0.0, 0.0, -shift]),
                    RevoluteJointBuilder::new(x).local_anchor2(point![-shift, 0.0, 0.0]),
                    RevoluteJointBuilder::new(z).local_anchor2(point![0.0, 0.0, -shift]),
                    RevoluteJointBuilder::new(x).local_anchor2(point![shift, 0.0, 0.0]),
                ];

                impulse_joints.insert(curr_parent, handles[0], revs[0], true);
                impulse_joints.insert(handles[0], handles[1], revs[1], true);
                impulse_joints.insert(handles[1], handles[2], revs[2], true);
                impulse_joints.insert(handles[2], handles[3], revs[3], true);

                curr_parent = handles[3];
            }
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints);
    testbed.look_at(point![478.0, 83.0, 228.0], point![134.0, 83.0, -116.0]);
}

File: python/rapier-testbed/rapier_testbed/examples3/stress_tests/joint_revolute3.py
"""Port of examples3d/stress_tests/joint_revolute3.rs."""
from __future__ import annotations

import rapier3d as rp
from ..._registry import register

CATEGORY = "Stress Tests"
NAME = "ImpulseJoint revolute"

X = (1.0, 0.0, 0.0)
Z = (0.0, 0.0, 1.0)


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    rad = 0.4
    num = 10
    shift = 2.0

    for l in range(4):
        y = l * shift * num * 3.0
        for j in range(50):
            x = j * shift * 4.0
            curr_parent = bodies.insert(rp.RigidBody.fixed().translation((x, y, 0.0)))
            colliders.insert_with_parent(rp.Collider.cuboid(rad, rad, rad), curr_parent, bodies)
            for i in range(num):
                z = i * shift * 2.0 + shift
                positions = [
                    (x, y, z),
                    (x + shift, y, z),
                    (x + shift, y, z + shift),
                    (x, y, z + shift),
                ]
                handles = []
                for k in range(4):
                    body = rp.RigidBody.dynamic().translation(positions[k])
                    h = bodies.insert(body)
                    handles.append(h)
                    colliders.insert_with_parent(
                        rp.Collider.cuboid(rad, rad, rad).density(1.0), h, bodies
                    )

                revs = [
                    rp.RevoluteJoint.builder(axis=Z).local_anchor2((0.0, 0.0, -shift)).build(),
                    rp.RevoluteJoint.builder(axis=X).local_anchor2((-shift, 0.0, 0.0)).build(),
                    rp.RevoluteJoint.builder(axis=Z).local_anchor2((0.0, 0.0, -shift)).build(),
                    rp.RevoluteJoint.builder(axis=X).local_anchor2((shift, 0.0, 0.0)).build(),
                ]
                impulse_joints.insert(curr_parent, handles[0], revs[0], wake_up=True)
                impulse_joints.insert(handles[0], handles[1], revs[1], wake_up=True)
                impulse_joints.insert(handles[1], handles[2], revs[2], wake_up=True)
                impulse_joints.insert(handles[2], handles[3], revs[3], wake_up=True)
                curr_parent = handles[3]

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.look_at((478.0, 83.0, 228.0), (134.0, 83.0, -116.0))


register(CATEGORY, NAME, init_world, dim=3)


if __name__ == "__main__":
    from ..._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
